### PR TITLE
Use character columns instead of UTF-8 columns in the diagnostics printer

### LINKED ISSUE #2507: `DiagnosticsFormatter` crashes when source code contains an emoji
### Description

`DiagnosticsFormatter` crashes with a  `Fatal error: Range requires lowerBound <= upperBound` message when the source code contains an emoji. 

Screenshot 2024-02-21 at 10 51 44 PM

### Steps to Reproduce

Run: 

```bash
swift run swift-parser-cli print-diags -s 'func o() {
  _ = {
    0
    /x}}1️⃣} /
    2
  }
}'
```

This version:
```bash
swift run swift-parser-cli print-diags -s 'func o() {
  _ = {
    0
    /x}}$} /
    2
  }
}'
```
works.

Tested on `main`.

### PR DESCRIPTION
**Pull request by ahoppen**
Fixes a crash and improves the printing of errors on lines that contain multi-byte UTF-8 characters.

Fixes #2507
rdar://123409748

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please test

**Comment by ahoppen** [comment]
@swift-ci Please test

File: Sources/SwiftDiagnostics/DiagnosticsFormatter.swift [hunk 1]
```diff
@@ -68,6 +68,22 @@ public struct DiagnosticsFormatter {
...
       return diagnostics.isEmpty && suffixText.isEmpty
     }
+
+    /// Converts a UTF-8 column index into an index that considers each character as a single column, not each UTF-8
+    /// byte.
+    ///
+    /// For example the 👨‍👩‍👧‍👦 character is considered as a single character, not 25 bytes.
+    ///
+    /// Both the input and the output column are 1-based.
+    func characterColumn(ofUtf8Column utf8Column: Int) -> Int {
```
Comment (bnbarham) [inline Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:78]: I assume you don't think this should be part of the location converter?

**Review by bnbarham — APPROVED** [review]

File: Sources/SwiftDiagnostics/DiagnosticsFormatter.swift (hunk 1 above)
Comment (ahoppen) [reply Sources/SwiftDiagnostics/DiagnosticsFormatter.swift:78]: No, I think it’s quite specific to this use case.

**Comment by ahoppen** [comment]
@swift-ci Please test Windows

**Review by Matejkob — COMMENTED** [review]
I'm not so sure if considering emojis as one character is correct. Are they two-character symbols? 

I believe in this test:
```swift
func testEmojiInSourceCode() {
    let source = """
      let 👨‍👩‍👧‍👦 = ;
      """

    let expectedOutput = """
      1 │ let 👨‍👩‍👧‍👦 = ;
        │         ╰─ error: expected expression in variable
      """

    assertStringsEqualWithDiff(annotate(source: source), expectedOutput)
}
```
the diagnostic should point to `;` not the space between `=` and `;`. I can not test it right now, but I think if we change the variable name to regular characters like so:
```swift
func testEmojiInSourceCode() {
    let source = """
      let e = ;
      """

    let expectedOutput = """
      1 │ let e = ;
        │         ╰─ error: expected expression in variable
      """

    assertStringsEqualWithDiff(annotate(source: source), expectedOutput)
}
```
the annotation will be pointing to semi column.

**Comment by ahoppen** [comment]
The problem is that, as far as I know, it depends on the font that renders the font. A font can render an emoji as a single-width character, a double width character, anything in between or however it likes, really. And as far as I know there’s no Unicode specification for it and my thought that assuming that 👨‍👩‍👧‍👦 is rendered as a single character is closer to the truth than assuming that it gets rendered as 25 characters.

**Comment by Matejkob** [comment]
Ran the emoji test, and indeed, the annotation shifts on difficult Xcode font (also in the terminal), highlighting the complexity of emoji rendering in code. 
Screenshot 2024-02-27 at 9 09 49 PM
Your explanation about the unpredictable rendering of emojis due to font differences is super insightful, thanks!

Out of curiosity, I've built this C code:
```c
#include <stdio.h>

int main() {
    printf("Hello, World!🤯")
    return 0;
}
```
and the annotation in the diagnostic method produced by gcc is shifted as well:
 Screenshot 2024-02-27 at 9 11 48 PM

**Comment by al45tair** [comment]
> The problem is that, as far as I know, it depends on the font that renders the font. A font can render an emoji as a single-width character, a double width character, anything in between or however it likes, really. And as far as I know there’s no Unicode specification for it and my thought that assuming that 👨‍👩‍👧‍👦 is rendered as a single character is closer to the truth than assuming that it gets rendered as 25 characters.

This is mainly to do with terminal emulator behaviour rather than the font (though the font can certainly make things worse if it so desires); generally speaking terminal emulators try to match the behaviour of the `wcwidth()` function on their respective platform, and _if_ that supports Unicode, it tends to just use the narrow or wide property from the East Asian Width table without doing anything much more complicated. That usually goes wrong for cases where the base character is narrow but some Emoji sequence is in play (so that the result is wide).

**Pull request merged**